javascript/javascript2/week2/index.js

```diff
@@ -0,0 +1,125 @@
+
+
+// 1. Movies with a short title 
+
+const shortMovies = movies.filter(short => short.title.length <= 3);
+console.log(shortMovies);
+
+
+
+// 2. Movies with long title
+
+const longMovies = movies.filter(long => long.title.length >=20 );
+console.log(longMovies);
+
+
+
+// 3. Count the number of movies between 1980-1989 including both 
+
+const count = movies.filter(title => {
+    return title.year >= 1980 && title.year <= 1989;
+}).length;
+console.log(count);
+
+
+
+// 4. Add new tag key based on rating
+
+const addTagKey = movies.map(function(addKey) {
+    if(addKey.rating >= 7) {
+        addKey.tag = "Good";
+    }
+    else if(addKey.rating >= 4 && addKey.rating < 7) {
+        addKey.tag = "Average";
+    }
+    else {
+        addKey.tag = "Bad";
+    }
+    return addKey;
+});
+console.log(addTagKey);
+
+
+
+// 5. Using chaining method
+
+const filteredMovies = movies.filter(title => title.rating > 6)
+.map(title => title.rating);
+console.log(filteredMovies);
+
+
+
+// 6. Count the total number of movies based on keywords: Surfer, Alien or Benjamin
+
+const countKeyWords = movies.reduce((counter, movie) => {
+    if (movie.title.toLowerCase().includes ("surfer") ||
+        movie.title.toLowerCase().includes ("alien") ||
+        movie.title.toLowerCase().includes ("benjamin") )
+
+    counter += 1;
+    
+    return counter;
+  }, 0); 
+  
+console.log(countKeyWords);
+
+
+
+
+// 7. Create duplicate array
+
+const titlesWithDuplicateWords = [];
+
+movies.map(movie => movie.title.split(" "))
+	.forEach(titleWord =>
+		titleWord.filter((word, index) =>
+			titleWord.indexOf(word) != index
+				? titlesWithDuplicateWords.push(titleWord)
+				: false
+		)
+	);
```

Comment:
Very inventive solution! Beautiful work of method chaining here, the gods of functional programming are surely smiling from above! 
 
However something is off about the logic, and I'm not entirely sure what it is. For one thing, we're not account for differences purely due to capitalization,so if you add `movie.title.toLowerCase().split(" ")` instead you end up with 236 titles. A number of those are also counted twice! 
 
Something to keep in mind is the intended usage of `map`, `filter` and `forEach`. `map` and `filter` are functions which take arrays as input and also output arrays, so it's good to think of them as singular "transformations" of an input array - and here we're using both methods as essentially for-loops, without really using the output. 
 
Here's my solution, which finds 204 movies: 
```js
const duplicatedTitles = movies.filter(movie => {
    const wordCount = {};
    movie.title.toLowerCase()
        .split(" ")
        .forEach(w => wordCount[w] = !wordCount[w] ? 1 : wordCount[w] + 1);
    return Object.keys(wordCount).some(key => wordCount[key] > 1);
});
```

Here I'm finding movies by filtering on the following way: for each movie title, keep a word count object that maps each word to an integer count. As we loop over the words in the (lowercased) title with `forEach`, we update the word count object by either initializing the word to 1, or incrementing it if it's already there. Then I return true if any of the words have a value greater than 1.